homeassistant/components/kodi/binary_sensor.py

```diff
@@ -0,0 +1,165 @@
+"""Sensor classes for kodi."""
+from __future__ import annotations
+
+import logging
+
+from jsonrpc_base.jsonrpc import TransportError
+
+from homeassistant.components.binary_sensor import BinarySensorEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DATA_CONNECTION, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Kodi sensors based on a config entry."""
+    data = hass.data[DOMAIN][config_entry.entry_id]
+    connman = data[DATA_CONNECTION]
+    if (uid := config_entry.unique_id) is None:
+        uid = config_entry.entry_id
+
+    sensor_entities = [
+        KodiBinaryEntity(
+            "Screensaver",
+            "GUI.OnScreensaverActivated",
+            "GUI.OnScreensaverDeactivated",
+            "System.ScreenSaverActive",
+            uid,
+            connman,
+        ),
+        KodiBinaryEntity(
+            "Energy saving",
+            "GUI.OnDPMSActivated",
+            "GUI.OnDPMSDeactivated",
+            "System.DPMSActive",
+            uid,
+            connman,
+        ),
+    ]
+    async_add_entities(sensor_entities)
+
+
+class KodiBinaryEntity(BinarySensorEntity):
+    """Generic binary sensor entity for WebSocket callbacks."""
+
+    _attr_has_entity_name = True
+
+    def __init__(self, name, api_on, api_off, boolean, uid, connman):
+        """Initialize kodi binary sensor."""
+        self._is_on = False
+        self._attr_name = name
+        self._attr_unique_id = uid + "_" + api_on
+        self._api_on = api_on
+        self._api_off = api_off
+        self._kodi_boolean = boolean
+        self._connman = connman
+        self._kodi_uid = uid
+
+    @property
+    def is_on(self):
+        """Return the state of the sensor."""
+        return self._is_on
+
+    @property
+    def device_info(self):
+        """Return kodi uid to register entity with the kodi device."""
+        return DeviceInfo(
+            identifiers={(DOMAIN, self._kodi_uid)},
+        )
+
+    @callback
+    def async_on(self, sender, data):  # pylint: disable=unused-argument
+        """Switch sensor on from api call."""
+        self._is_on = True
+        self.async_write_ha_state()
+        _LOGGER.debug("Kodi %s on (%s)", self.name, self._api_on)
+
+    @callback
+    def async_off(self, sender, data):  # pylint: disable=unused-argument
+        """Switch sensor off from api call."""
+        self._is_on = False
+        self.async_write_ha_state()
+        _LOGGER.debug("Kodi %s off (%s)", self.name, self._api_off)
+
+    @callback
+    def async_dummy(self, sender, data):
```

Comment:
Is there really no better way to remove the callbacks?

Reply:
Well, I looked at the [source](https://github.com/emlove/jsonrpc-websocket) for the websocket server and from what I read something like `server.Player.OnPause` is implemented using nested namespaces?! I'm not "Python" enough to understand that to see if there is some kind of "clear()" function available. Do you have a clue about that?

Reply:
TBH this is code that was already there when I started tinkering with the Kodi integration, and I definitely don't know the innards of the library. Might be worth pinging the maintainer of that library